collection_encoders.go

```diff
@@ -0,0 +1,196 @@
+		return err
+	}
+	_, err = s.encoder.writer.SeekPos(s.lengthPos)
+	if err != nil {
+		return err
+	}
+	err = s.encoder.writer.WriteLengthUint64(uint64(s.length))
+	if err != nil {
+		return err
+	}
+	_, err = s.encoder.writer.SeekPos(finalPos)
+	s.encoder.begin.Store(false)
+	return err
+}
+
+func (s *baseCollectionEncoder) WriteFieldStrStr(key string, value string) error {
+	panic("implement me")
+}
+
+func (s *baseCollectionEncoder) WriteFieldStr(field string) error {
+	panic("implement me")
+}
+
+func (s *baseCollectionEncoder) WriteFieldStrFloat64(field string, value float64) error {
+	panic("implement me")
+}
+
+func (s *baseCollectionEncoder) WriteFieldStrStrWithExpiry(key string, value string, expiry *time.Time) error {
+	panic("implement me")
+}
+
+type ListEncoder struct {
+	baseCollectionEncoder
+}
+
+func NewListEncoder(e *Encoder) (*ListEncoder, error) {
+	encoder := &ListEncoder{}
+	encoder.encoder = e
+	err := encoder.WriteZeroLength()
+	return encoder, err
+}
+
+func (s *ListEncoder) WriteFieldStr(val string) error {
+	s.length++
+	return s.encoder.writeString(val)
+}
+
+type SetEncoder struct {
+	baseCollectionEncoder
+}
+
+func NewSetEncoder(e *Encoder) (*SetEncoder, error) {
+	encoder := &SetEncoder{}
+	encoder.encoder = e
+	err := encoder.WriteZeroLength()
+	if err != nil {
+		return nil, err
+	}
+	return encoder, nil
+}
+
+func (s *SetEncoder) WriteFieldStr(field string) error {
+	s.length++
+	return s.encoder.writeString(field)
+}
+
+type SortedSetEncoder struct {
+	baseCollectionEncoder
+}
+
+func NewSortedSetEncoder(e *Encoder) (*SortedSetEncoder, error) {
+	encoder := &SortedSetEncoder{}
+	encoder.encoder = e
+	err := encoder.WriteZeroLength()
+	if err != nil {
+		return nil, err
+	}
+	return encoder, nil
+}
+
+func (s *SortedSetEncoder) WriteFieldStrFloat64(key string, value float64) error {
+	err := s.encoder.writeString(key)
+	if err != nil {
+		return err
+	}
+	score := math.Float64bits(value)
+	err = s.encoder.writer.WriteUint64(score)
+	if err != nil {
+		return err
+	}
+	s.length++
+	return nil
+}
+
+type HashEncoder struct {
+	baseCollectionEncoder
+}
+
+func NewHashEncoder(e *Encoder) (*HashEncoder, error) {
+	encoder := &HashEncoder{}
+	encoder.encoder = e
+	err := encoder.WriteZeroLength()
+	if err != nil {
+		return nil, err
+	}
+	return encoder, nil
+}
+
+func (s *HashEncoder) WriteFieldStrStr(key string, value string) error {
+	err := s.encoder.writeString(key)
+	if err != nil {
+		return err
+	}
+	err = s.encoder.writeString(value)
+	if err != nil {
+		return err
+	}
+	s.length++
+	return nil
+}
+
+type HashMetadataEncoder struct {
+	baseCollectionEncoder
+}
+
+func NewHashMetadataEncoder(e *Encoder) (*HashMetadataEncoder, error) {
+	encoder := &HashMetadataEncoder{}
+	encoder.encoder = e
+	err := encoder.WriteZeroLength()
+	if err != nil {
+		return nil, err
+	}
+	return encoder, nil
+}
+
+func (s *HashMetadataEncoder) WriteFieldStrStrWithExpiry(key string, value string, expiry *time.Time) error {
```

Comment:
minor: is there a reason why expiry is a pointer rather than a value? Documentation of suggests using values, and its zero value having unix time of zero should play well with the code we have. 

https://pkg.go.dev/time#Time
```
// Programs using times should typically store and pass them as values,
// not pointers. That is, time variables and struct fields should be of
// type [time.Time], not *time.Time.
```